src/NServiceBus.Core/Performance/Metrics/ReceiveDiagnosticsBehavior.cs

```diff
@@ -0,0 +1,63 @@
+namespace NServiceBus
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics.Metrics;
+    using System.Threading.Tasks;
+    using Pipeline;
+
+    class ReceiveDiagnosticsBehavior : IBehavior<IIncomingPhysicalMessageContext, IIncomingPhysicalMessageContext>
+    {
+        static readonly Meter NServiceBusMeter = new Meter(
+            NServiceBusDiagnosticsInfo.InstrumentationName,
+            NServiceBusDiagnosticsInfo.InstrumentationVersion);
+
+        static readonly Counter<long> TotalProcessedSuccessfully =
+            NServiceBusMeter.CreateCounter<long>("messaging.successes", description: "Total number of messages processed successfully by the endpoint.");
+
+        static readonly Counter<long> TotalFetched =
+            NServiceBusMeter.CreateCounter<long>("messaging.fetches", description: "Total number of messages fetched from the queue by the endpoint.");
+
+        static readonly Counter<long> TotalFailures =
+            NServiceBusMeter.CreateCounter<long>("messaging.failures", description: "Total number of messages processed unsuccessfully by the endpoint.");
+
+        public ReceiveDiagnosticsBehavior(string endpointName, string queueNameBase, string discriminator)
+        {
+            this.endpointName = endpointName;
+            this.queueNameBase = queueNameBase;
+            this.discriminator = discriminator ?? "";
+        }
+
+        public async Task Invoke(IIncomingPhysicalMessageContext context, Func<IIncomingPhysicalMessageContext, Task> next)
+        {
+            context.MessageHeaders.TryGetMessageType(out var messageTypes);
+
+            var tags = new List<KeyValuePair<string, object>>
+            {
+                new KeyValuePair<string, object>("messaging.endpoint", endpointName),
+                new KeyValuePair<string, object>("messaging.discriminator", discriminator),
+                new KeyValuePair<string, object>("messaging.queue", queueNameBase),
+                new KeyValuePair<string, object>("messaging.type", messageTypes),
+            };
```

Comment:
Call `.ToArray()` here to prevent calling it twice. If these are the only tags we're expecting then create an array directly.

It may be optimization overkill but the first 3 are static once the endpoint starts so we could new those up once and pass the same instance to each tags array.

Reply:
There is the exception and message type to add as well. As the array changes, not sure how much optimization this would result in.